File: daily_python/2311_Longest_Binary_Subsequence_Less_Than_or_Equal_to_K.py
import os
import sys
from typing import List

class Solution:
    def longestSubsequence(self, s: str, k: int) -> int:
        ans = sum(1 if c == '0' else 0 for c in s)
        curr = 0
        for i in range(len(s)):
            c = s[len(s) - 1 - i]
            if c == '0':
                continue
            else:
                curr += (1 << i)
                if curr > k:
                    break
                ans += 1
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            s = lines[0].split('=')[1].strip()[1:-2]
            k = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((s, k))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for s, k in dataset:
        results.append(solution.longestSubsequence(s, k))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2025/2311_Longest_Binary_Subsequence_Less_Than_or_Equal_to_K.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    int longestSubsequence(string s, int k) {
        int n = s.length();
        int ans = 0;
        for (char c : s) if (c == '0') ans++;
        int curr = 0;
        int pos = 1;
        for (int i = n - 1; i >= 0; i--){
            if (s[i] == '1'){
                curr += pos;
                if (curr > k) break;
                ans++;
            }
            if (pos > 1e9) break;
            pos *= 2;

        }
        return ans;
    }
};

int main(){
    string s = "1001010";
    int k = 5;
    // string s = "00101001";
    // int k = 1;
    // string s = "000101010011011001011101111000111111100001011000000100010000111100000011111001000111100111101001111001011101001011011101001011011001111111010011100011110111010000010000010111001001111101100001111";
    // int k = 300429827;
    Solution S;
    cout << S.longestSubsequence(s, k) << endl;
    return 0;
}